### ISSUE TITLE
SQLClient connection leak in getConnectionAwait(), leads to a deadlock

### ISSUE THREAD

**Issue by ph4r05** [vert-x3/vertx-lang-kotlin#195]
### Version

- vertx 3.9.7
- kotlin 1.5.0 (tested also on 1.4.32)
- open-jdk-14 

### Context

I observed a resource leak in JDBCClient, namely in the `vertx-lang-kotlin:3.9.7` function

```kotlin
suspend fun SQLClient.getConnectionAwait(): SQLConnection 
```

The problem is if coroutine calling `getConnectionAwait()` is cancelled before the handler on the following method is called:

```kotlin
SQLClient getConnection(Handler<AsyncResult<SQLConnection>> handler);
```

If such situation happens (e.g., calling REST connection is terminated / timed out), the obtained connection, passed to the handler, is not closed and returned to the pool. This causes leakage, depleting all free connections (if using with connection pool like C3P0). The program ends up in the terminal state without usable DB connection.

I believe that the problem lies in the https://github.com/vert-x3/vertx-lang-kotlin/blob/d991b1683c05c6d5081ee2f6513838c9c1dc977c/vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/VertxCoroutine.kt#L63

```kotlin
suspend fun <T> awaitEvent(block: (h: Handler<T>) -> Unit): T {
  return suspendCancellableCoroutine { cont: CancellableContinuation<T> ->
    try {
      block.invoke(Handler { t ->
        cont.resume(t) // <---- here, if cont is cancelled, t remains opened
      })
    } catch (e: Exception) {
      cont.resumeWithException(e)
    }
  }
}
```

I solved the problem by implementing the coroutine bridging on my own with a if-branch handling cancelled coroutine, such as

```kotlin
    try {
      block.invoke(Handler { t ->
        if (cont.isActive)
          cont.resume(t)
        } else {
          t.close()  // <--- close connection if coroutine is already cancelled when handler is invoked
        } 
      })
    } catch (e: Exception) {
      cont.resumeWithException(e)
    }
```

I noticed that this method is deprecated https://github.com/vert-x3/vertx-jdbc-client/issues/196 https://github.com/vert-x3/vertx-lang-kotlin/blob/c1938f890b711734d3e191d274422d3f3098726a/vertx-lang-kotlin/src/main/kotlin/io/vertx/kotlin/ext/sql/SQLClient.kt#L64-L69 , but this may help others having the same issues if they didn't migrate yet. Btw was it deprecated also because of this potential issue?

### Do you have a reproducer?

Not yet, I can create one if description is not enough.

### Steps to reproduce

1. setup MySQL JDBC-backed demo
2. spawn 2000 worker threads simulating clients, each querying the database (code below), let clients timeout randomly
3. watch number of failures / timeouts

```kotlin
val rnd = Random()
withTimeout(rnd.nextLong().rem(1000L)) {
    dbClient.getConnectionAwait().use {
        it.queryAwait("SELECT 1")
    }
}
```

**Comment by vietj** [comment]
can you provide a reproducer project ? that would help

**Comment by ph4r05** [comment]
@vietj I hoped that my description was detailed enough to avoid implementing PoC as I am quite busy 😅 
 
Here is the POC:
https://github.com/ph4r05/vertx-sqlconn-poc

Check out readme for the usage. There is also a fix implemented.

If you cannot reproduce it, increase number of clients from 500 to 2000 and dos-time from 10_000 to 20_000.

Code of interest:
https://github.com/ph4r05/vertx-sqlconn-poc/blob/659e211be57ffabde712fa750c5b5dd36f30ccde/src/main/kotlin/me/deadcode/vertx/WebServer.kt#L106-L112

Buggy ping
https://github.com/ph4r05/vertx-sqlconn-poc/blob/659e211be57ffabde712fa750c5b5dd36f30ccde/src/main/kotlin/me/deadcode/vertx/WebServer.kt#L129-L133

Ping with fixed race condition
https://github.com/ph4r05/vertx-sqlconn-poc/blob/659e211be57ffabde712fa750c5b5dd36f30ccde/src/main/kotlin/me/deadcode/vertx/WebServer.kt#L135-L179

**Comment by vietj** [comment]
can you elaborate on how the coroutine can get cancelled in your case ?

**Comment by vietj** [comment]
I understand the issue and it seems it is related to the fact that Vert.x futures cannot be cancelled and I am wondering how we could solve this. e.g we could a close called when the coroutine is cancelled and the future result implements an interface like Closeable.

**Comment by vietj** [comment]
Perhaps also we could make Vert.x coroutines as non cancellable ?

**Comment by ph4r05** [comment]
> can you elaborate on how the coroutine can get cancelled in your case ?

Coroutine gets usually cancelled due to timeout in `withTimeout()` as we want to limit execution time of some code paths, not to lock too many resources if server is overloaded (e.g., if DB server is overloaded or has opened maximum amount of connections) and to respond in timely manner.

Another reason is client disconnect during the request, e.g., in the websocket server handler. The coroutine chain triggered by the client request get cancelled. Similarly, the CoroutineScope may be cancelled from another reason (e.g., reloading TLS certificates, maintenance mode).

> e.g we could a close called when the coroutine is cancelled and the future result implements an interface like Closeable.

I was also thinking along these lines. It seems like a fine workaround. It `t` comes from the Future, it is probably sound to close it if coroutine is not active anymore. 

> Perhaps we could make Vert.x coroutines as non cancellable ?

I am not sure about the impact of this. Personally, it feels a bit risky and maybe too restrictive.

**Comment by vietj** [comment]
I would actually discourage using such timeout. The more I think about it, the more I see coroutine cancellation as something similar to killing a thread without having an opportunity to cleanly close all the resources associated with the coroutines.

Instead you should try to use timeout on the resources you are using and let them fail for you, e.g if your HTTP server uses a WebClient then set timeout on the HTTP requests. It will have a similar effect and has a more reliable handling.

**Comment by ph4r05** [comment]
I actually quite like structured concurrency principles coroutines offer. As cancellation is cooperative and it throws an exception, it is possible to reasonably close resources with 'use' or 'try-finally'. However, bridging coroutines with Future callbacks are a bit more tricky with respect to cancellation. https://medium.com/androiddevelopers/cancellation-in-coroutines-aa6b90163629

Good point with the resources timeouts. Using timeout also on the top level of coroutine processing makes sure we didn't forget setting any timeouts and guarantee maximal execution time for a request (availability, scalability).

**Comment by vietj** [comment]
no, we are saying that we will not implement cancelation as this is not what we want to support, e.g supporting loom in the future will not be able to deal with cancelation
